src/calliope/backend/backend_model.py

```diff
@@ -269,10 +299,8 @@ def _add_component(
         """
         references: set[str] = set()
 
-        if component_dict is None:
-            component_dict = self.inputs.math[component_type][name]
-        if name not in self.inputs.math[component_type]:
-            self.inputs.math[component_type][name] = component_dict
+        if name not in self.inputs.math.get(component_type, {}):
```

Comment:
no, you can simply do `model.backend.add_constraint(...)` and it will add a new constraint to the backend _after_ you've added everything from the math dict. This then appends it to the math dict.

Reply:
Hmmm... this suggests that we are not arranging logic in the best way. If that condition can only be reached when calling `add_x`, then this should be handled at that level.

Maybe setting a new `math` key should happen in `add_x` (keeping the override off, to catch duplicates), and this function should just add the component to the backend?

Reply:
Except then you have to handle in every `add_x` method separately, including across backends (something this `add_component` helper func was made to deal with). It could be added as a wrapper on all `add_x` funcs, I guess?

Reply:
Hmmm... I do not think that is the case, because `component_type` and `component_dict` are passed by the caller, but repeating it everywhere seems off, I agree. He probably have an antipattern.

Let's keep it for now, as this is another issue.